csrc/scheduler/normalization_inner.cpp

```diff
@@ -203,19 +204,29 @@ int64_t getMaxRegisterCountPerThread(
       register_overhead;
 
   // (1) use register calculated from target occupancy
-  int64_t register_per_thread_target = getRegPerThreadGivenThreadsPerSM(
-      target_blocks_per_sm * threads_per_block);
+  int64_t threads_per_sm =
+      std::min(threads_per_block * target_blocks_per_sm, max_threads_per_sm);
```

Comment:
I think `threads_per_sm = std::min(threads_per_block * target_blocks_per_sm, max_threads_per_sm);` guarantees `threads_per_sm` won't beyond the allowed resident threads per SM.

Reply:
sorry I was reading this wrong earlier and didn't realize here we are just rounding up the target_threads_per_sm.

Looking at it here, we are basically doing:
```
int64_t threads_per_sm = 
ceilDiv(target_warps_per_sm * threads_per_warp, threads_per_block) * threads_per_block;
```
And then followed by `threads_per_sm = std::min(threads_per_sm, max_threads_per_sm);`

The cap here with `max_threads_per_sm` doesn't look right. Should that be a round down to a multiple of `threads_per_block` instead?

Reply:
make sense! I moved up the check:
```
  // ensure higher than target by round up
  int64_t target_blocks_per_sm =
      ceilDiv(target_warps_per_sm * threads_per_warp, threads_per_block);
  // ensure lower than hardware limit by round down
  if(threads_per_block * target_blocks_per_sm > max_threads_per_sm){
    target_blocks_per_sm = max_threads_per_sm / threads_per_block;
  }
```

This change won't influence the perf because our current `target_warps_per_sm * threads_per_warp = 28 * 32 = 896`, the worst case is round up from 895 threads to 1790 threads which is smaller than `max_threads_per_sm`

Reply:
I prefer to have it here since 8.6~8.9 has an odd 1536 cap... :vomiting_face:

Reply:
new knowledge!